### PR TITLE
feat(deployment): expand macros in S3 options

### PR DESCRIPTION
**Pull request by piotrdubiel**
S3 options can now contain macros to allow something like this:
```
"publish": {
    "provider": "s3",
    "bucket": "some-bucket",
    "path": "${os}/${env.SOME_ENV}"
}
```
Now `bucket`, `path` and `channel` can contain macros. Previously these options couldn't be configured with macros.

### CODE REVIEW

**Comment by develar** [comment]
Resolving in this place is ok, because in any case result is not cached — to be able to specify publish config per target. Anyway, arch is not used currently.

File: packages/electron-builder/src/publish/PublishManager.ts [hunk 1]
```diff
@@ -380,7 +380,33 @@ export async function getPublishConfigs(packager: PlatformPackager<any>, targetS
...
+  return await (<Promise<Array<PublishConfiguration>>>BluebirdPromise.map(asArray(publishers), it => getResolvedPublishConfig(packager.info, typeof it === "string" ? {provider: it} : it)))
+    .then(publishConfigs => expandPublishConfigs(packager, publishConfigs))
+}
+
+function expandPublishConfigs(packager: PlatformPackager<any>, publishConfigs: Array<PublishConfiguration>) {
+  return publishConfigs.map(publishConfig => expandPublishConfig(packager, publishConfig))
+}
+
+function expandPublishConfig(packager: PlatformPackager<any>, publishConfig: PublishConfiguration) {
+  if ((<S3Options>publishConfig).bucket != null) {
```
Comment (develar) [inline packages/electron-builder/src/publish/PublishManager.ts]: I think we can just expand macro for all keys in the object. It also will allow us to remove explicit expand for generic URL (see usages in this file). Could you please do it?

Comment (piotrdubiel) [inline packages/electron-builder/src/publish/PublishManager.ts]: Done. One question: is it ok not to add arch in `computeDownloadUrl` to expand?

Comment (piotrdubiel) [inline packages/electron-builder/src/publish/PublishManager.ts]: Too soon, need to fix.

Comment (develar) [inline packages/electron-builder/src/publish/PublishManager.ts]: Only snap test should be failed (unrelated).

Comment (piotrdubiel) [inline packages/electron-builder/src/publish/PublishManager.ts]: I think it's fixed.

**Pull request merged**